lib/live_cluster/manage_controller.py

```
@@ -3411,3 +3412,251 @@ async def _do_default(self, line):
+
+        if self.warn and not self.prompt_challenge(
+            f"You're about to add a masking rule for bin '{bin_name}' in namespace '{namespace}', set '{set_name}' with function '{function_str}'."
+        ):
+            return
+
+        info_resp = await self.cluster.info_masking_add_rule(
+            namespace,
+            set_name,
+            bin_name,
+            bin_type,
+            function_name,
+            function_params,
+            nodes="principal",
+        )
+
+        resp = list(info_resp.values())[0]
+
+        if isinstance(resp, Exception):
+            raise resp
+
+        self.view.print_result(
+            "Successfully added masking rule. Use 'show masking' to view the rules."
+        )
+
+    def _parse_function_params(self, line):
+        """Parse function parameters dynamically."""
+        function_params = {}
+        reserved_keywords = {
+            "namespace",
+            "set",
+            "bin",
+            "type",
+        }  # Use set for O(1) lookup
+
+        # Count parameters first for validation (single pass)
+        param_count = 0
+        for i, arg in enumerate(line):
+            if arg in reserved_keywords:
+                break
+            param_count += 1
+
+        # Ensure even number of arguments for proper key-value pairing
+        if param_count % 2 != 0:
+            param_args = line[:param_count]
+            raise ShellException(
+                f"Function parameters must be in key-value pairs. Found {param_count} arguments: {' '.join(param_args)}"
+            )
+
+        # Parse parameters in pairs (single pass, no copying)
+        while line and line[0] not in reserved_keywords:
+            if len(line) >= 2:
+                param_name = line.pop(0)
+                param_value = line.pop(0)
+                function_params[param_name] = param_value
+            else:
+                break  # Should not happen due to validation above
```

If this is an unexpected failure state, lets raise an exception or return some error here

---

I don't see any changes here

---

I now see the changes